### ISSUE TITLE
[BUG] "Requiring @playwright/test second time" when using shared package

### ISSUE THREAD

**Issue by Forfold** [microsoft/playwright#15819]
**Code Snippet**

https://github.com/Forfold/playwright-test

**Describe the bug**

Related to #15349 and #11889

I have a `playwright-auth` package I'm creating to handle some login stuff that I plan to use in multiple repos I'm working on. `playwright-auth` exports a `login()` and a `globalLogin()` function. I've created the above repo `playwright-test` acting as my auth package.

I see `Error: Requiring @playwright/test second time` but both repos utilize `@playwright/test` for different reasons.

My `playwright-auth` package utilizes `chromium`, `expect`, `request`, and `Page`.

My website repos use `test` and `expect`

Full error output:
```
Running 0 test using 0 worker


Error: Requiring @playwright/test second time, 
First:
Error: 

   at node_modules/@playwright/test/lib/index.js:73

  71 |   }
  72 | } else {
> 73 |   process['__pw_initiator__'] = new Error().stack;
     |                                 ^
  74 | }
  75 |
  76 | const test = _baseTest.extend({

    at Object.<anonymous> (/Users/Z0027PF/git/playwright-test/example/node_modules/@playwright/test/lib/index.js:73:33)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1174:10)
    at Module.load (node:internal/modules/cjs/loader:998:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at Module.require (node:internal/modules/cjs/loader:1022:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/Z0027PF/git/playwright-test/example/node_modules/@playwright/test/index.js:17:13)
    at Module._compile (node:internal/modules/cjs/loader:1120:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1174:10)
    at Module.load (node:internal/modules/cjs/loader:998:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at Module.require (node:internal/modules/cjs/loader:1022:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/Z0027PF/git/playwright-test/example/playwright/login.spec.ts:1:1)

Second: 
    at Object.<anonymous> (/Users/Z0027PF/git/playwright-test/example/node_modules/playwright-test/node_modules/@playwright/test/lib/index.js:68:11)
    at Object.<anonymous> (/Users/Z0027PF/git/playwright-test/example/node_modules/playwright-test/node_modules/@playwright/test/index.js:17:13)
    at Object.<anonymous> (/Users/Z0027PF/git/playwright-test/example/node_modules/playwright-test/dist/index.js:16:16)
    at Object.<anonymous> (/Users/Z0027PF/git/playwright-test/example/playwright/login.spec.ts:5:23)

=================
 no tests found.
=================
```

**Comment by dgozman** [comment]
@Forfold This is because you use a relative link between two modules, and each has a copy of `@playwright/test` in their own `node_modules`. Take a look at [this comment](https://github.com/microsoft/playwright/issues/11817#issuecomment-1043630558) for the solution. Let me know if this helps.

**Comment by Forfold** [comment]
I have my `playwright-auth` package published on our internal artifactory. I used a relative path for purposes of the example so I wouldn't have to go through publishing a dummy package to npm

**Comment by dgozman** [comment]
I did the following and everything works.

```bash
git clone git@github.com:Forfold/playwright-test.git
cd playwright-test
npm install
npx tsc -p .
npm pack
cd example
# playwright.config.ts seems to be in the wrong place, based on the relative paths it uses
mv src/playwright.config.ts .
# make sure to install "playwright-test" helper as a real package, not a relative link
npm install ../playwright-test-0.1.0.tgz
npx playwright test
```

There are no errors and the single test successfully prints `appPage`:

```
Running 1 test using 1 worker
[WebServer] (node:77725) [DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE] DeprecationWarning: 'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
(Use `node --trace-deprecation ...` to show where the warning was created)
[WebServer] (node:77725) [DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE] DeprecationWarning: 'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.

[chromium] › login.spec.ts:5:3 › test login using playwright-test › login
<ref *1> Page {
  ...
}

  1 passed (4s)

To open last HTML report run:

  npx playwright show-report
```

**Comment by Forfold** [comment]
but how would this work when my package is installed via my internal npm registry? that is the issue im facing.

**Comment by dgozman** [comment]
@Forfold Sorry, I don't have an internal npm registry to quickly check that, but I'd assume it's exactly the same as installing from a packed package. Note that in your example you don't have `@playwright/test` as a prod dependency in the helper package, which is the right way to do it.

**Comment by dgozman** [comment]
@Forfold We have to reproduce this issue to be able to debug it. If you can provide a full repro that we can replicate locally, please file a new issue with it, and link to this one.

**Comment by Forfold** [comment]
> # playwright.config.ts seems to be in the wrong place, based on the relative paths it uses
> mv src/playwright.config.ts .

From your comment it looks like I may need a second config file for playwright in my repo. The config included there is for the example website only.


However, regarding installing the package from a tgz file, I don't know if I would expect my users to do that. I would assume that importing `login` should work out of the box if they install my `playwright-auth` package and have their own playwright config setup.

I'll do some more exhaustive testing locally. I don't really have it in me to host a dummy package on NPM just for this example

**Comment by RomarioNijimAlmond** [comment]
im facing the issue myself and all I have are simple files, pages and tests, both files imports playwright test but I can't seem to run the test because of this error

page:
import { expect } from "@playwright/test";
import { BasePage } from "./BasePage";

export class HomePage extends BasePage {

    public async validateBody(text: string) {
        const bodyTag = document.body.innerText
        expect(bodyTag).toContain(text);
    }
}

testfile:

import { test } from '@playwright/test';
import { BasePage } from '../pages/BasePage';
import { HomePage } from '../pages/HomePage';


test.describe('test', async () => {
    let basePage: BasePage;
    let homePage: HomePage;

    test.beforeEach(async ({ page }) => {
        basePage = new BasePage(page);
        homePage = new HomePage(page);
    })

    test.afterEach(async ({ context }) => {
        await context.clearCookies();
    })
    
    test.only('validate body text', async() => {
        await test.step('validate body', async() => {
            await basePage.loadApplication('http://localhost:3000/');
            await homePage.validateBody('hello world');
        })
    })
})

**Comment by msieorbus** [comment]
I still have same issue, i have two separate npm packages in our internal feed, both needs to have @playwright/test import, im installing them later in my tests project and still have this error.

**Comment by pfteter** [comment]
This issue can be solved with hoisting in npm workspaces. npm handles hoistes @playwright/test

**Comment by maksnester** [comment]
I have several packages in a monorepo that have playwright as a dependency and I started to get this error after the minor version playwright update.

```
$ npm ls @playwright/test
hm
├─┬ @hm/e2e-tools@0.0.0 -> ./hm-tests/e2e-tools
│ └── @playwright/test@1.50.1 <----- not deduped
└─┬ @hm/e2e@0.0.0 -> ./hm-tests/e2e
  ├─┬ @forsta/playwright@0.0.4
  │ └── @playwright/test@1.50.1 deduped
  └── @playwright/test@1.50.1  <----- not deduped
```

I had to run `"npm exec --workspaces -c 'rm -rf ./node_modules' && rm -rf ./node_modules && rm ./package-lock.json && npm i",`

then it got correctly deduplicated and the error went away

```
hm
├─┬ @hm/e2e-tools@0.0.0 -> ./hm-tests/e2e-tools
│ └── @playwright/test@1.50.1
└─┬ @hm/e2e@0.0.0 -> ./hm-tests/e2e
  ├─┬ @forsta/playwright@0.0.4
  │ └── @playwright/test@1.50.1 deduped
  └── @playwright/test@1.50.1 deduped
```

**Comment by fantonangeli** [comment]
In my case VSCode installed a `node_modules` inside a test folder of a package in my monorepo.
To solve it I needed to run:
`rm -r tests-e2e/node_modules tests-e2e/package-lock.json` 
from inside the monorepo's package.

**Comment by mstandiferpaylocity** [comment]
Know this is closed, but this may help put someone in the right direction, if none of items here listed work for you. I just came across error when I upgraded to 1.55.0. My issue turned out to be, cause by a typo, it was in my fixture, an import for playwright was like this  'playwright/test, instead of '@playwright/test', and since none of the clearing of node_modules, package-lock.json worked. I looked a little closer at the error messages, and in a log, buried was a reference to a fixture ( the errors from the basic log output were very misleading, nothing about the fixture ), as soon as I opened it, it was underlined in red for the import. Oddly, I had not touched this fixture in quite a while, it was one merged into a main fixture, and it had been working. But I was always having an issue in the Test Explorer, and selecting different playwright configs I have, they would show up with duplicated entries all the time. Once I fixed that import that stopped happening.
